File: backend/app/schemas/participation.py
"""Pydantic schemas for the shared Participation Workspace domain (Project /
Training / Workshop), database/migrations/062-065_participation_*.sql.

ONE reusable domain, not three duplicated ones -- every model here is
discriminated by `kind` ('PROJECT' | 'TRAINING' | 'WORKSHOP') rather than
having a separate schema module per opportunity type, mirroring how
`applications.opportunity_type` already discriminates INTERNSHIP/JOB rows
in one table (app.schemas.application).

Field names mirror the migrations exactly. `industry_id` is never accepted
from a client anywhere in this module -- ownership is always derived
server-side (from the opportunity for programs, from the application for
workspaces), matching every other opportunity-owned table in this schema.
"""

from typing import Literal

from pydantic import BaseModel, ConfigDict, Field, model_validator

ParticipationKind = Literal["PROJECT", "TRAINING", "WORKSHOP"]
ProgramStatus = Literal["DRAFT", "PUBLISHED", "ARCHIVED"]
ResourceType = Literal["VIDEO", "DOCUMENT", "LINK", "REFERENCE", "OTHER"]
WorkspaceStatus = Literal["ACTIVE", "COMPLETED"]
SubmissionStatus = Literal["SUBMITTED", "UNDER_REVIEW", "REVIEWED"]
ReviewStatus = Literal["REVIEWED", "NEEDS_REVISION", "ACCEPTED"]
FeedbackType = Literal["GENERAL", "PROGRESS", "STRENGTH", "IMPROVEMENT", "FINAL_NOTE"]
RecommendedLevel = Literal["Beginner", "Intermediate", "Advanced", "Expert"]
RecommendationPriority = Literal["LOW", "MEDIUM", "HIGH"]
EvaluationStatus = Literal["DRAFT", "FINALIZED"]
CompletionStatus = Literal["COMPLETED"]


def _blank_to_none(data: object) -> object:
    if not isinstance(data, dict):
        return data
    cleaned: dict = {}
    for key, value in data.items():
        if isinstance(value, str):
            value = value.strip() or None
        cleaned[key] = value
    return cleaned


# ---- programs ----


class ProgramCreate(BaseModel):
    """POST body. Exactly one of project_id/training_id/workshop_id must be
    supplied, matching `kind` -- the route re-validates ownership of that
    opportunity before the insert; `industry_id` is never part of this
    model."""

    model_config = ConfigDict(extra="forbid")

    kind: ParticipationKind
    project_id: str | None = None
    training_id: str | None = None
    workshop_id: str | None = None
    title: str = Field(min_length=1, max_length=200)
    description: str | None = Field(default=None, max_length=10_000)

    @model_validator(mode="before")
    @classmethod
    def _normalise(cls, data: object) -> object:
        return _blank_to_none(data)

    @model_validator(mode="after")
    def _one_opportunity(self) -> "ProgramCreate":
        picked = {"PROJECT": self.project_id, "TRAINING": self.training_id, "WORKSHOP": self.workshop_id}
        if not picked[self.kind]:
            raise ValueError(f"{self.kind.lower()}_id is required for kind={self.kind}.")
        others = [v for k, v in picked.items() if k != self.kind]
        if any(others):
            raise ValueError("Only the opportunity id matching `kind` may be set.")
        return self


class ProgramUpdate(BaseModel):
    model_config = ConfigDict(extra="forbid")

    title: str | None = Field(default=None, min_length=1, max_length=200)
    description: str | None = Field(default=None, max_length=10_000)

    @model_validator(mode="before")
    @classmethod
    def _normalise(cls, data: object) -> object:
        return _blank_to_none(data)


class ProgramOpportunityRef(BaseModel):
    id: str
    title: str
    status: str


class ProgramResponse(BaseModel):
    id: str
    kind: ParticipationKind
    project_id: str | None = None
    training_id: str | None = None
    workshop_id: str | None = None
    title: str
    description: str | None = None
    status: ProgramStatus
    published_at: str | None = None
    created_at: str | None = None
    updated_at: str | None = None
    opportunity: ProgramOpportunityRef | None = None


class ProgramListResponse(BaseModel):
    programs: list[ProgramResponse]


# ---- modules ----


class ModuleCreate(BaseModel):
    model_config = ConfigDict(extra="forbid")

    title: str = Field(min_length=1, max_length=200)
    description: str | None = Field(default=None, max_length=5_000)
    sequence_order: int = Field(default=0, ge=0)

    @model_validator(mode="before")
    @classmethod
    def _normalise(cls, data: object) -> object:
        return _blank_to_none(data)


class ModuleUpdate(BaseModel):
    model_config = ConfigDict(extra="forbid")

    title: str | None = Field(default=None, min_length=1, max_length=200)
    description: str | None = Field(default=None, max_length=5_000)
    sequence_order: int | None = Field(default=None, ge=0)

    @model_validator(mode="before")
    @classmethod
    def _normalise(cls, data: object) -> object:
        return _blank_to_none(data)


class ModuleResponse(BaseModel):
    id: str
    program_id: str
    title: str
    description: str | None = None
    sequence_order: int
    is_published: bool
    created_at: str | None = None
    updated_at: str | None = None


class ModuleListResponse(BaseModel):
    modules: list[ModuleResponse]


# ---- resources ----


class ResourceCreate(BaseModel):
    model_config = ConfigDict(extra="forbid")

    module_id: str | None = None
    title: str = Field(min_length=1, max_length=200)
    description: str | None = Field(default=None, max_length=5_000)
    resource_type: ResourceType
    resource_url: str | None = Field(default=None, max_length=2_000)
    sequence_order: int = Field(default=0, ge=0)

    @model_validator(mode="before")
    @classmethod
    def _normalise(cls, data: object) -> object:
        return _blank_to_none(data)


class ResourceUpdate(BaseModel):
    model_config = ConfigDict(extra="forbid")

    title: str | None = Field(default=None, min_length=1, max_length=200)
    description: str | None = Field(default=None, max_length=5_000)
    resource_url: str | None = Field(default=None, max_length=2_000)
    sequence_order: int | None = Field(default=None, ge=0)

    @model_validator(mode="before")
    @classmethod
    def _normalise(cls, data: object) -> object:
        return _blank_to_none(data)


class ResourceResponse(BaseModel):
    id: str
    program_id: str
    module_id: str | None = None
    title: str
    description: str | None = None
    resource_type: ResourceType
    resource_url: str | None = None
    sequence_order: int
    is_published: bool
    created_at: str | None = None


class ResourceListResponse(BaseModel):
    resources: list[ResourceResponse]


# ---- assignments ----


class AssignmentCreate(BaseModel):
    model_config = ConfigDict(extra="forbid")

    module_id: str | None = None
    title: str = Field(min_length=1, max_length=200)
    description: str | None = Field(default=None, max_length=5_000)
    instructions: str | None = Field(default=None, max_length=10_000)
    due_at: str | None = None
    max_score: float | None = Field(default=None, gt=0)
    is_required: bool = True

    @model_validator(mode="before")
    @classmethod
    def _normalise(cls, data: object) -> object:
        return _blank_to_none(data)


class AssignmentUpdate(BaseModel):
    model_config = ConfigDict(extra="forbid")

    title: str | None = Field(default=None, min_length=1, max_length=200)
    description: str | None = Field(default=None, max_length=5_000)
    instructions: str | None = Field(default=None, max_length=10_000)
    due_at: str | None = None
    max_score: float | None = Field(default=None, gt=0)
    is_required: bool | None = None

    @model_validator(mode="before")
    @classmethod
    def _normalise(cls, data: object) -> object:
        return _blank_to_none(data)


class AssignmentResponse(BaseModel):
    id: str
    program_id: str
    module_id: str | None = None
    title: str
    description: str | None = None
    instructions: str | None = None
    due_at: str | None = None
    max_score: float | None = None
    is_required: bool
    is_published: bool
    sequence_order: int
    created_at: str | None = None
    updated_at: str | None = None


class AssignmentListResponse(BaseModel):
    assignments: list[AssignmentResponse]


# ---- workspaces ----


class WorkspaceOpportunityRef(BaseModel):
    id: str
    title: str
    status: str


class WorkspaceResponse(BaseModel):
    id: str
    kind: ParticipationKind
    student_id: str
    student_name: str | None = None
    industry_id: str
    project_id: str | None = None
    training_id: str | None = None
    workshop_id: str | None = None
    project_application_id: str | None = None
    training_application_id: str | None = None
    workshop_application_id: str | None = None
    workspace_status: WorkspaceStatus
    started_at: str | None = None
    completed_at: str | None = None
    created_at: str | None = None
    updated_at: str | None = None
    opportunity: WorkspaceOpportunityRef | None = None
    program_id: str | None = None
    progress: "ProgressSummary | None" = None


class WorkspaceListResponse(BaseModel):
    workspaces: list[WorkspaceResponse]


class EnsureWorkspaceRequest(BaseModel):
    """POST body for the idempotent ensure-or-create operation. Exactly one
    of the three application ids must be supplied, matching `kind`."""

    model_config = ConfigDict(extra="forbid")

    kind: ParticipationKind
    project_application_id: str | None = None
    training_application_id: str | None = None
    workshop_application_id: str | None = None

    @model_validator(mode="after")
    def _one_application(self) -> "EnsureWorkspaceRequest":
        picked = {
            "PROJECT": self.project_application_id,
            "TRAINING": self.training_application_id,
            "WORKSHOP": self.workshop_application_id,
        }
        if not picked[self.kind]:
            raise ValueError(f"{self.kind.lower()}_application_id is required for kind={self.kind}.")
        others = [v for k, v in picked.items() if k != self.kind]
        if any(others):
            raise ValueError("Only the application id matching `kind` may be set.")
        return self


class ProgressSummary(BaseModel):
    completed_required: int
    published_required: int
    percent: int  # 0-100, 0 when published_required == 0


# ---- submissions ----


class SubmissionCreate(BaseModel):
    model_config = ConfigDict(extra="forbid")

    assignment_id: str
    submission_text: str | None = Field(default=None, max_length=20_000)
    submission_url: str | None = Field(default=None, max_length=2_000)

    @model_validator(mode="before")
    @classmethod
    def _normalise(cls, data: object) -> object:
        return _blank_to_none(data)

    @model_validator(mode="after")
    def _has_content(self) -> "SubmissionCreate":
        if not self.submission_text and not self.submission_url:
            raise ValueError("Provide submission_text and/or submission_url.")
        return self


class SubmissionResponse(BaseModel):
    id: str
    workspace_id: str
    assignment_id: str
    assignment_title: str | None = None
    attempt_number: int
    submission_text: str | None = None
    submission_url: str | None = None
    submitted_at: str | None = None
    status: SubmissionStatus
    created_at: str | None = None
    latest_review: "ReviewResponse | None" = None


class SubmissionListResponse(BaseModel):
    submissions: list[SubmissionResponse]


# ---- reviews ----


class ReviewCreate(BaseModel):
    model_config = ConfigDict(extra="forbid")

    score: float | None = Field(default=None, ge=0)
    feedback: str | None = Field(default=None, max_length=10_000)
    status: ReviewStatus

    @model_validator(mode="before")
    @classmethod
    def _normalise(cls, data: object) -> object:
        return _blank_to_none(data)


class ReviewResponse(BaseModel):
    id: str
    submission_id: str
    reviewer_id: str
    score: float | None = None
    feedback: str | None = None
    status: ReviewStatus
    reviewed_at: str | None = None
    created_at: str | None = None


class ReviewListResponse(BaseModel):
    reviews: list[ReviewResponse]


# ---- feedback ----


class FeedbackCreate(BaseModel):
    model_config = ConfigDict(extra="forbid")

    feedback_type: FeedbackType
    title: str | None = Field(default=None, max_length=200)
    feedback: str = Field(min_length=1, max_length=10_000)

    @model_validator(mode="before")
    @classmethod
    def _normalise(cls, data: object) -> object:
        return _blank_to_none(data)


class FeedbackResponse(BaseModel):
    id: str
    workspace_id: str
    industry_id: str
    feedback_type: FeedbackType
    title: str | None = None
    feedback: str
    created_at: str | None = None


class FeedbackListResponse(BaseModel):
    feedback: list[FeedbackResponse]


# ---- skill recommendations ----


class SkillRecommendationCreate(BaseModel):
    model_config = ConfigDict(extra="forbid")

    skill_id: str
    recommended_level: RecommendedLevel | None = None
    reason: str = Field(min_length=1, max_length=2_000)
    priority: RecommendationPriority = "MEDIUM"

    @model_validator(mode="before")
    @classmethod
    def _normalise(cls, data: object) -> object:
        return _blank_to_none(data)


class SkillRecommendationResponse(BaseModel):
    id: str
    workspace_id: str
    skill_id: str
    skill_name: str | None = None
    recommended_level: RecommendedLevel | None = None
    reason: str
    priority: RecommendationPriority
    created_by: str
    created_at: str | None = None


class SkillRecommendationListResponse(BaseModel):
    recommendations: list[SkillRecommendationResponse]


# ---- evaluation criteria ----


class CriterionCreate(BaseModel):
    model_config = ConfigDict(extra="forbid")

    name: str = Field(min_length=1, max_length=200)
    description: str | None = Field(default=None, max_length=2_000)
    max_score: float = Field(gt=0)
    weight: float = Field(gt=0, le=100)
    sequence_order: int = Field(default=0, ge=0)

    @model_validator(mode="before")
    @classmethod
    def _normalise(cls, data: object) -> object:
        return _blank_to_none(data)


class CriterionUpdate(BaseModel):
    model_config = ConfigDict(extra="forbid")

    name: str | None = Field(default=None, min_length=1, max_length=200)
    description: str | None = Field(default=None, max_length=2_000)
    max_score: float | None = Field(default=None, gt=0)
    weight: float | None = Field(default=None, gt=0, le=100)
    sequence_order: int | None = Field(default=None, ge=0)

    @model_validator(mode="before")
    @classmethod
    def _normalise(cls, data: object) -> object:
        return _blank_to_none(data)


class CriterionResponse(BaseModel):
    id: str
    program_id: str
    name: str
    description: str | None = None
    max_score: float
    weight: float
    sequence_order: int
    created_at: str | None = None
    updated_at: str | None = None


class CriterionListResponse(BaseModel):
    criteria: list[CriterionResponse]


# ---- evaluations ----


class EvaluationScoreInput(BaseModel):
    model_config = ConfigDict(extra="forbid")

    criterion_id: str
    score: float = Field(ge=0)
    feedback: str | None = Field(default=None, max_length=2_000)

    @model_validator(mode="before")
    @classmethod
    def _normalise(cls, data: object) -> object:
        return _blank_to_none(data)


class EvaluationScoreResponse(BaseModel):
    id: str
    evaluation_id: str
    criterion_id: str
    criterion_name: str | None = None
    max_score: float | None = None
    score: float
    feedback: str | None = None


class EvaluationSaveRequest(BaseModel):
    """PUT body for saving a DRAFT evaluation's scores/feedback. Rejected
    once the evaluation is FINALIZED (409, not a validation error --
    the route checks current status first)."""

    model_config = ConfigDict(extra="forbid")

    overall_feedback: str | None = Field(default=None, max_length=10_000)
    scores: list[EvaluationScoreInput] = Field(default_factory=list)

    @model_validator(mode="before")
    @classmethod
    def _normalise(cls, data: object) -> object:
        return _blank_to_none(data)


class EvaluationResponse(BaseModel):
    id: str
    workspace_id: str
    evaluator_id: str
    status: EvaluationStatus
    overall_score: float | None = None
    overall_feedback: str | None = None
    evaluated_at: str | None = None
    created_at: str | None = None
    updated_at: str | None = None
    scores: list[EvaluationScoreResponse] = Field(default_factory=list)


# ---- completion ----


class CompletionResponse(BaseModel):
    id: str
    workspace_id: str
    final_evaluation_id: str | None = None
    completed_at: str | None = None
    completion_status: CompletionStatus
    final_score: float | None = None
    created_at: str | None = None

File: frontend/types/participation.ts
// Mirrors backend/app/schemas/participation.py and
// database/migrations/062-066_participation_*.sql. ONE reusable domain
// for PROJECT/TRAINING/WORKSHOP, discriminated by `kind` -- never three
// duplicated type modules.

export const PARTICIPATION_KINDS = ["PROJECT", "TRAINING", "WORKSHOP"] as const;
export type ParticipationKind = (typeof PARTICIPATION_KINDS)[number];

export const PARTICIPATION_KIND_LABELS: Record<ParticipationKind, string> = {
  PROJECT: "Project",
  TRAINING: "Training",
  WORKSHOP: "Workshop",
};

export type ProgramStatus = "DRAFT" | "PUBLISHED" | "ARCHIVED";
export type ResourceType = "VIDEO" | "DOCUMENT" | "LINK" | "REFERENCE" | "OTHER";
export const RESOURCE_TYPES: ResourceType[] = ["VIDEO", "DOCUMENT", "LINK", "REFERENCE", "OTHER"];

export type WorkspaceStatus = "ACTIVE" | "COMPLETED";
export type SubmissionStatus = "SUBMITTED" | "UNDER_REVIEW" | "REVIEWED";
export const SUBMISSION_STATUS_LABELS: Record<SubmissionStatus, string> = {
  SUBMITTED: "Submitted",
  UNDER_REVIEW: "Under Review",
  REVIEWED: "Reviewed",
};
export type ReviewStatus = "REVIEWED" | "NEEDS_REVISION" | "ACCEPTED";
export const REVIEW_STATUSES: ReviewStatus[] = ["ACCEPTED", "NEEDS_REVISION", "REVIEWED"];
export const REVIEW_STATUS_LABELS: Record<ReviewStatus, string> = {
  ACCEPTED: "Accepted",
  NEEDS_REVISION: "Needs revision",
  REVIEWED: "Reviewed",
};

export type FeedbackType = "GENERAL" | "PROGRESS" | "STRENGTH" | "IMPROVEMENT" | "FINAL_NOTE";
export const FEEDBACK_TYPES: FeedbackType[] = ["GENERAL", "PROGRESS", "STRENGTH", "IMPROVEMENT", "FINAL_NOTE"];
export const FEEDBACK_TYPE_LABELS: Record<FeedbackType, string> = {
  GENERAL: "General",
  PROGRESS: "Progress",
  STRENGTH: "Strength",
  IMPROVEMENT: "Improvement",
  FINAL_NOTE: "Final note",
};

export type RecommendedLevel = "Beginner" | "Intermediate" | "Advanced" | "Expert";
export const RECOMMENDED_LEVELS: RecommendedLevel[] = ["Beginner", "Intermediate", "Advanced", "Expert"];

export type RecommendationPriority = "LOW" | "MEDIUM" | "HIGH";
export const RECOMMENDATION_PRIORITIES: RecommendationPriority[] = ["LOW", "MEDIUM", "HIGH"];

export type EvaluationStatus = "DRAFT" | "FINALIZED";

export interface OpportunityRef {
  id: string;
  title: string;
  status: string;
}

export interface ProgramResponse {
  id: string;
  kind: ParticipationKind;
  project_id: string | null;
  training_id: string | null;
  workshop_id: string | null;
  title: string;
  description: string | null;
  status: ProgramStatus;
  published_at: string | null;
  created_at: string | null;
  updated_at: string | null;
  opportunity: OpportunityRef | null;
}

export interface ModuleResponse {
  id: string;
  program_id: string;
  title: string;
  description: string | null;
  sequence_order: number;
  is_published: boolean;
  created_at: string | null;
  updated_at: string | null;
}

export interface ResourceResponse {
  id: string;
  program_id: string;
  module_id: string | null;
  title: string;
  description: string | null;
  resource_type: ResourceType;
  resource_url: string | null;
  sequence_order: number;
  is_published: boolean;
  created_at: string | null;
}

export interface AssignmentResponse {
  id: string;
  program_id: string;
  module_id: string | null;
  title: string;
  description: string | null;
  instructions: string | null;
  due_at: string | null;
  max_score: number | null;
  is_required: boolean;
  is_published: boolean;
  sequence_order: number;
  created_at: string | null;
  updated_at: string | null;
}

export interface ProgressSummary {
  completed_required: number;
  published_required: number;
  percent: number;
}

export interface WorkspaceResponse {
  id: string;
  kind: ParticipationKind;
  student_id: string;
  student_name: string | null;
  industry_id: string;
  project_id: string | null;
  training_id: string | null;
  workshop_id: string | null;
  project_application_id: string | null;
  training_application_id: string | null;
  workshop_application_id: string | null;
  workspace_status: WorkspaceStatus;
  started_at: string | null;
  completed_at: string | null;
  created_at: string | null;
  updated_at: string | null;
  opportunity: OpportunityRef | null;
  program_id: string | null;
  progress: ProgressSummary | null;
}

export interface ReviewResponse {
  id: string;
  submission_id: string;
  reviewer_id: string;
  score: number | null;
  feedback: string | null;
  status: ReviewStatus;
  reviewed_at: string | null;
  created_at: string | null;
}

export interface SubmissionResponse {
  id: string;
  workspace_id: string;
  assignment_id: string;
  assignment_title: string | null;
  attempt_number: number;
  submission_text: string | null;
  submission_url: string | null;
  submitted_at: string | null;
  status: SubmissionStatus;
  created_at: string | null;
  latest_review: ReviewResponse | null;
}

export interface FeedbackResponse {
  id: string;
  workspace_id: string;
  industry_id: string;
  feedback_type: FeedbackType;
  title: string | null;
  feedback: string;
  created_at: string | null;
}

export interface SkillRecommendationResponse {
  id: string;
  workspace_id: string;
  skill_id: string;
  skill_name: string | null;
  recommended_level: RecommendedLevel | null;
  reason: string;
  priority: RecommendationPriority;
  created_by: string;
  created_at: string | null;
}

export interface CriterionResponse {
  id: string;
  program_id: string;
  name: string;
  description: string | null;
  max_score: number;
  weight: number;
  sequence_order: number;
  created_at: string | null;
  updated_at: string | null;
}

export interface EvaluationScoreResponse {
  id: string;
  evaluation_id: string;
  criterion_id: string;
  criterion_name: string | null;
  max_score: number | null;
  score: number;
  feedback: string | null;
}

export interface EvaluationResponse {
  id: string;
  workspace_id: string;
  evaluator_id: string;
  status: EvaluationStatus;
  overall_score: number | null;
  overall_feedback: string | null;
  evaluated_at: string | null;
  created_at: string | null;
  updated_at: string | null;
  scores: EvaluationScoreResponse[];
}

export interface CompletionResponse {
  id: string;
  workspace_id: string;
  final_evaluation_id: string | null;
  completed_at: string | null;
  completion_status: "COMPLETED";
  final_score: number | null;
  created_at: string | null;
}

/** The three FK field names, keyed by kind -- used to build request
 * bodies generically instead of branching three times at every call
 * site. */
export const OPPORTUNITY_FK: Record<ParticipationKind, "project_id" | "training_id" | "workshop_id"> = {
  PROJECT: "project_id",
  TRAINING: "training_id",
  WORKSHOP: "workshop_id",
};

export const APPLICATION_FK: Record<
  ParticipationKind,
  "project_application_id" | "training_application_id" | "workshop_application_id"
> = {
  PROJECT: "project_application_id",
  TRAINING: "training_application_id",
  WORKSHOP: "workshop_application_id",
};

/** Route base path per kind, for cross-linking back to the opportunity's
 * own existing pages (never invented -- these match the routes already
 * shipped for Project/Training/Workshop). */
export const INDUSTRY_OPPORTUNITY_BASE_PATH: Record<ParticipationKind, string> = {
  PROJECT: "/industry/projects",
  TRAINING: "/industry/training",
  WORKSHOP: "/industry/workshops",
};

export const STUDENT_OPPORTUNITY_BASE_PATH: Record<ParticipationKind, string> = {
  PROJECT: "/student/industry-projects",
  TRAINING: "/student/trainings",
  WORKSHOP: "/student/workshops",
};
